### DISCUSSION TITLE
[FEATURE REQUEST] Allow macro addition to Button class

### DISCUSSION THREAD

**Discussion by martynaskre** [Power-Components/livewire-powergrid#718] [Ideas]
# Allow macro addition to Button class

## Summary

Currently, [Button](https://github.com/Power-Components/livewire-powergrid/blob/main/src/Button.php) class does not allow any modifications. Class cannot be extended nor it can be extended with macros.

### Why is this needed?

Let's say Powergrid is customised with custom template. Then, the template requires icons with actions buttons. Buttons cannot be hardcoded, because there are different types of them.

### Suggested Solution(s)

The most simple solution would be to add `Illuminate\Support\Traits\Macroable` trait to Button class. This would allow to easily extend class with required methods, e.g. `$button->icon('...')`. Of course it would still require some refactoring in [Helpers/Actions](https://github.com/Power-Components/livewire-powergrid/blob/main/src/Helpers/Actions.php) class, which should not be major.

**Comment by MarkasCherry** [comment]
Yes! This would be so useful!

**Comment by luanfreitasdev** [comment]
Hello. could you share more examples of how to use it and what we would need to refactor?

**Reply by martynaskre** [reply]
This [repository](https://github.com/martynaskre/powergrid-feature-example) contains sample project which contains following modifications:
- `Button` class now [implements macroable trait](https://github.com/martynaskre/powergrid-feature-example/blob/master/packages/livewire-powergrid/src/Button.php#L9) and has additional property named [dynamicProperties](https://github.com/martynaskre/powergrid-feature-example/blob/master/packages/livewire-powergrid/src/Button.php#L41).
- `Actions` class now also has [dynamicProperties](https://github.com/martynaskre/powergrid-feature-example/blob/master/packages/livewire-powergrid/src/Helpers/Actions.php#L44). Dynamic properties are copied when class is constructed. Also this class has method [getDynamicProperty](https://github.com/martynaskre/powergrid-feature-example/blob/master/packages/livewire-powergrid/src/Helpers/Actions.php#L327).

Dynamic properties can be declared in service providers, e.g. [here](https://github.com/martynaskre/powergrid-feature-example/blob/master/app/Providers/LivewirePowergridServiceProvider.php#L27). After dynamic property is declared, then it can be used as real method, e.g. [here](https://github.com/martynaskre/powergrid-feature-example/blob/master/app/Http/Livewire/UsersTable.php#L162).

All things considered, this would not require a lot of modifications and would not cause major code changes. If this solution is good, then I can submit a pull request.

**Reply by luanfreitasdev** [reply]
This is interesting, would it work as if it were a ? attribute. what else could we do with it?

**Reply by martynaskre** [reply]
What do you mean by saying *? attribute*?

Well, using macros on a button class allows end user to customise template and include custom button properties. There are other possibilities for macros, but this one is bothering me right now.

**Reply by luanfreitasdev** [reply]
Got it, we can work on this in a PR and add some tests with some real examples and add in the doc

**Reply by martynaskre** [reply]
I have submitted a PR #726.
